File: typescript/tests/gpu-accountant.test.ts
/**
 * Per-task GPU accountant — cgroup walk + NVML snapshot pair + dual-event emission.
 * Mirrors python/tests/test_gpu_accountant.py.
 *
 * The TS deviation is dependency injection: the GpuAccountant constructor
 * accepts an Options object with nvml/cgroup hooks (matching the gpu-runtime
 * pattern) so tests inject deterministic stubs.
 */

import { describe, it, expect, beforeEach } from "vitest";
import {
  GpuAccountant,
  _resetWarningStateForTests,
  type GpuAccountantHooks,
} from "../src/core/gpu-accountant.js";
import { GpuRuntimeKind } from "../src/core/gpu-runtime.js";
import type { CloudEnv } from "../src/cloud-detect.js";
import type { CgroupScope } from "../src/core/cgroup-walker.js";
import type { UtilSample, MemInfo } from "../src/core/nvml-reader.js";

function cloud(
  provider: string | null,
  region: string | null,
  source: "env" | "imds" | "dmi" | "none",
  instanceType?: string | null,
): CloudEnv {
  return { provider, region, source, instanceType: instanceType ?? null };
}

const SELF = process.pid;

function defaultMemInfo(): MemInfo {
  return { usedBytes: 21474836480, totalBytes: 85899345920 };
}

function baseHooks(
  overrides: Partial<GpuAccountantHooks> = {},
): GpuAccountantHooks {
  return {
    initNvml: () => true,
    getDeviceCount: () => 1,
    getProductName: (_i) => "nvidia h100 80gb hbm3",
    getMigMode: (_i) => false,
    getMemoryInfo: (_i) => defaultMemInfo(),
    classifyScope: (): CgroupScope => ({ kind: "container", path: "/docker/abc" }),
    enumeratePids: (_scope) => [SELF],
    getProcessUtilization: (_i, _ts) => ({}),
    ...overrides,
  };
}

describe("GpuAccountant — Modal serverless emission", () => {
  beforeEach(() => _resetWarningStateForTests());

  it("emits one gpu_cost + one gpu_utilization_signal per device", () => {
    let call = 0;
    const samples: Array<Record<number, UtilSample>> = [
      {}, // baseline
      {
        [SELF]: {
          pid: SELF,
          smUtil: 80,
          memUtil: 30,
          timeStamp: 1_234_000,
        },
      },
    ];
    const hooks = baseHooks({
      getProcessUtilization: (_i, _ts) => samples[call++] ?? {},
    });
    const acc = new GpuAccountant(
      GpuRuntimeKind.Modal,
      cloud("modal", null, "env"),
      hooks,
    );
    acc.snapshotStart();
    const { costDetails, signalEvents } = acc.snapshotEndAndBuild(1234);
    expect(costDetails).not.toBeNull();
    expect(costDetails!.billing_model).toBe("per_gpu_second_active");
    expect(costDetails!.gpu_vendor).toBe("nvidia");
    expect(costDetails!.gpu_sku).not.toBeNull();
    expect(costDetails!.gpu_count).toBe(1);
    expect(costDetails!.duration_ms).toBe(1234);
    expect(costDetails!.cost_pending).toBe(true);
    expect(costDetails!.mig_profile).toBeNull();

    expect(signalEvents).not.toBeNull();
    expect(signalEvents!.length).toBe(1);
    const sig = signalEvents![0];
    expect(sig.gpu_index).toBe(0);
    expect(sig.gpu_sku).toBe(costDetails!.gpu_sku);
    expect(sig.sm_util_pct).not.toBeNull();
    expect(sig.vram_total_bytes).toBe(85899345920);
  });
});

describe("GpuAccountant — idempotency", () => {
  it("second snapshotEndAndBuild returns (null, null)", () => {
    _resetWarningStateForTests();
    const hooks = baseHooks({
      getDeviceCount: () => 0,
    });
    const acc = new GpuAccountant(
      GpuRuntimeKind.Modal,
      cloud("modal", null, "env"),
      hooks,
    );
    acc.snapshotStart();
    acc.snapshotEndAndBuild(1000);
    const second = acc.snapshotEndAndBuild(2000);
    expect(second.costDetails).toBeNull();
    expect(second.signalEvents).toBeNull();
  });
});

describe("GpuAccountant — no NVML / 0 devices → nothing", () => {
  it("init_nvml false → empty events", () => {
    _resetWarningStateForTests();
    const hooks = baseHooks({ initNvml: () => false });
    const acc = new GpuAccountant(
      GpuRuntimeKind.Modal,
      cloud("modal", null, "env"),
      hooks,
    );
    acc.snapshotStart();
    const r = acc.snapshotEndAndBuild(1000);
    expect(r.costDetails).toBeNull();
    expect(r.signalEvents).toBeNull();
  });

  it("0 devices → empty events", () => {
    _resetWarningStateForTests();
    const hooks = baseHooks({ getDeviceCount: () => 0 });
    const acc = new GpuAccountant(
      GpuRuntimeKind.AwsEc2Gpu,
      cloud("aws", "us-east-1", "imds", "p5.48xlarge"),
      hooks,
    );
    acc.snapshotStart();
    const r = acc.snapshotEndAndBuild(60_000);
    expect(r.costDetails).toBeNull();
    expect(r.signalEvents).toBeNull();
  });
});

describe("GpuAccountant — Decision #1 fallback labels", () => {
  it("bare_metal_user_slice → _cgroup_scope_fallback=no_container_scope", () => {
    _resetWarningStateForTests();
    let call = 0;
    const samples: Array<Record<number, UtilSample>> = [
      {},
      {
        [SELF]: { pid: SELF, smUtil: 50, memUtil: 20, timeStamp: 500_000 },
      },
    ];
    const hooks = baseHooks({
      classifyScope: () => ({ kind: "bare_metal_user_slice", path: null }),
      getMemoryInfo: () => ({ usedBytes: 0, totalBytes: 85899345920 }),
      getProcessUtilization: () => samples[call++] ?? {},
    });
    const acc = new GpuAccountant(
      GpuRuntimeKind.AwsEc2Gpu,
      cloud("aws", "us-east-1", "imds", "p5.48xlarge"),
      hooks,
    );
    acc.snapshotStart();
    const { costDetails } = acc.snapshotEndAndBuild(1000);
    expect(costDetails!._cgroup_scope_fallback).toBe("no_container_scope");
  });

  it("cgroup walk denied → _cgroup_scope_fallback=self_pid_only", () => {
    _resetWarningStateForTests();
    const hooks = baseHooks({
      getMemoryInfo: () => ({ usedBytes: 0, totalBytes: 80 * 1024 ** 3 }),
      enumeratePids: () => null,
      getProcessUtilization: () => ({}),
    });
    const acc = new GpuAccountant(
      GpuRuntimeKind.AwsEc2Gpu,
      cloud("aws", "us-east-1", "imds", "p5.48xlarge"),
      hooks,
    );
    acc.snapshotStart();
    const { costDetails } = acc.snapshotEndAndBuild(1000);
    expect(costDetails!._cgroup_scope_fallback).toBe("self_pid_only");
  });
});

describe("GpuAccountant — Decision #2 MIG transparency", () => {
  it("MIG detected → mig_profile populated, log fires", () => {
    _resetWarningStateForTests();
    const hooks = baseHooks({
      getProductName: () => "nvidia a100 80gb",
      getMigMode: () => true,
      getMemoryInfo: () => ({ usedBytes: 0, totalBytes: 80 * 1024 ** 3 }),
      getProcessUtilization: () => ({}),
    });
    const acc = new GpuAccountant(
      GpuRuntimeKind.AwsEc2Gpu,
      cloud("aws", "us-east-1", "imds", "p4d.24xlarge"),
      hooks,
    );
    acc.snapshotStart();
    const { costDetails } = acc.snapshotEndAndBuild(1000);
    expect(costDetails).not.toBeNull();
    expect(costDetails!.mig_profile).not.toBeNull();
  });
});

describe("GpuAccountant — Decision #3 window-averaged sm_util_pct", () => {
  it("4s@80% + 1s@0% over 5s → sm_util_pct ≈ 64%", () => {
    _resetWarningStateForTests();
    // Simulate accumulated active-GPU-microseconds = 3.2s (0.8 × 4_000_000)
    let call = 0;
    const samples: Array<Record<number, UtilSample>> = [
      {},
      {
        [SELF]: { pid: SELF, smUtil: 0, memUtil: 0, timeStamp: 3_200_000 },
      },
    ];
    const hooks = baseHooks({
      getMemoryInfo: () => ({ usedBytes: 0, totalBytes: 80 * 1024 ** 3 }),
      getProcessUtilization: () => samples[call++] ?? {},
    });
    const acc = new GpuAccountant(
      GpuRuntimeKind.Modal,
      cloud("modal", null, "env"),
      hooks,
    );
    acc.snapshotStart();
    const { costDetails, signalEvents } = acc.snapshotEndAndBuild(5000);
    expect(costDetails!.gpu_seconds_used).toBeGreaterThanOrEqual(3.0);
    expect(costDetails!.gpu_seconds_used).toBeLessThanOrEqual(3.4);
    expect(signalEvents![0].sm_util_pct).not.toBeNull();
    expect(signalEvents![0].sm_util_pct as number).toBeGreaterThanOrEqual(60);
    expect(signalEvents![0].sm_util_pct as number).toBeLessThanOrEqual(70);
  });
});

describe("GpuAccountant — sub-100ms degenerate", () => {
  it("duration_ms=0 → sm_util_pct = null (no div-by-zero)", () => {
    _resetWarningStateForTests();
    const hooks = baseHooks({
      getMemoryInfo: () => ({ usedBytes: 0, totalBytes: 80 * 1024 ** 3 }),
      getProcessUtilization: () => ({}),
    });
    const acc = new GpuAccountant(
      GpuRuntimeKind.Modal,
      cloud("modal", null, "env"),
      hooks,
    );
    acc.snapshotStart();
    const { signalEvents } = acc.snapshotEndAndBuild(0);
    expect(signalEvents).not.toBeNull();
    expect(signalEvents![0].sm_util_pct).toBeNull();
  });
});

describe("GpuAccountant — multi-device", () => {
  it("4 devices → 4 signal events with distinct gpu_index", () => {
    _resetWarningStateForTests();
    let utilCall = 0;
    const hooks = baseHooks({
      getDeviceCount: () => 4,
      getMemoryInfo: () => ({ usedBytes: 1 * 1024 ** 3, totalBytes: 80 * 1024 ** 3 }),
      getProcessUtilization: () => {
        utilCall++;
        if (utilCall <= 4) return {};
        return {
          [SELF]: { pid: SELF, smUtil: 50, memUtil: 20, timeStamp: 500_000 },
        };
      },
    });
    const acc = new GpuAccountant(
      GpuRuntimeKind.AwsEc2Gpu,
      cloud("aws", "us-east-1", "imds", "p5.48xlarge"),
      hooks,
    );
    acc.snapshotStart();
    const { costDetails, signalEvents } = acc.snapshotEndAndBuild(1000);
    expect(costDetails).not.toBeNull();
    expect(costDetails!.gpu_count).toBe(4);
    expect(signalEvents!.length).toBe(4);
    signalEvents!.forEach((sig, i) => {
      expect(sig.gpu_index).toBe(i);
    });
  });
});

File: go/core/gpu_accountant_test.go
// Task 6 — Per-task GPU accountant tests. Mirrors python commit 0d47371.

package core

import (
	"os"
	"testing"

	"github.com/DexwoxBusiness/dexcost-go/cloud"
)

// gpuMockWithDevices returns a ready-to-use mock NVML backend with N
// devices and self-PID accumulating SM time. The start sample has
// timeStamp=0; the end sample has timeStamp=1_000_000 (one full second).
// Using os.Getpid() ensures the PID falls into the accountant's cgroup
// PID union (which on non-container hosts degrades to self-PID-only).
func gpuMockWithDevices(deviceCount int) *MockNVMLBackend {
	selfPID := os.Getpid()
	utilStart := map[int]map[int]NVMLUtilSample{}
	utilEnd := map[int]map[int]NVMLUtilSample{}
	productNames := map[int]string{}
	mem := map[int]NVMLMemInfo{}
	for i := 0; i < deviceCount; i++ {
		productNames[i] = "NVIDIA H100 80GB HBM3"
		utilStart[i] = map[int]NVMLUtilSample{
			selfPID: {PID: selfPID, SMUtil: 0, MemUtil: 0, TimeStamp: 0},
		}
		utilEnd[i] = map[int]NVMLUtilSample{
			selfPID: {PID: selfPID, SMUtil: 80, MemUtil: 30, TimeStamp: 1_000_000},
		}
		mem[i] = NVMLMemInfo{UsedBytes: 40 * 1024 * 1024 * 1024, TotalBytes: 80 * 1024 * 1024 * 1024}
	}
	return &MockNVMLBackend{
		Available:    true,
		DeviceCount:  deviceCount,
		ProductNames: productNames,
		Memory:       mem,
		PerCallUtilization: map[int][]map[int]NVMLUtilSample{
			0: {utilStart[0], utilEnd[0]},
		},
	}
}

func TestGpuAccountantSnapshotEndEmitsCostEvent(t *testing.T) {
	resetNVMLForTests()
	mock := gpuMockWithDevices(1)
	SetNVMLBackendForTests(mock.AsBackend())
	t.Cleanup(resetNVMLForTests)

	a := NewGpuAccountant(GpuRuntimeModal, cloud.CloudEnv{Provider: "modal"})
	a.SnapshotStart()
	cost, signals := a.SnapshotEndAndBuild(1000)

	if cost == nil {
		t.Fatalf("expected gpu_cost details; got nil")
	}
	if cost["billing_model"] != "per_gpu_second_active" {
		t.Errorf("billing_model = %v; want per_gpu_second_active", cost["billing_model"])
	}
	if cost["cost_pending"] != true {
		t.Errorf("cost_pending should be true on emit")
	}
	if cost["gpu_vendor"] != "nvidia" {
		t.Errorf("gpu_vendor = %v; want nvidia", cost["gpu_vendor"])
	}
	if cost["gpu_count"] != 1 {
		t.Errorf("gpu_count = %v; want 1", cost["gpu_count"])
	}
	if len(signals) != 1 {
		t.Errorf("expected 1 utilization-signal event; got %d", len(signals))
	}
}

func TestGpuAccountantIdempotentFinalize(t *testing.T) {
	resetNVMLForTests()
	mock := gpuMockWithDevices(1)
	SetNVMLBackendForTests(mock.AsBackend())
	t.Cleanup(resetNVMLForTests)

	a := NewGpuAccountant(GpuRuntimeModal, cloud.CloudEnv{Provider: "modal"})
	a.SnapshotStart()
	cost1, _ := a.SnapshotEndAndBuild(1000)
	cost2, signals2 := a.SnapshotEndAndBuild(1000)
	if cost1 == nil {
		t.Fatalf("first finalize returned nil")
	}
	if cost2 != nil || signals2 != nil {
		t.Errorf("idempotent second call should return nil/nil; got %v / %v", cost2, signals2)
	}
}

func TestGpuAccountantNoNVMLReturnsNil(t *testing.T) {
	resetNVMLForTests() // noop backend
	t.Cleanup(resetNVMLForTests)
	a := NewGpuAccountant(GpuRuntimeModal, cloud.CloudEnv{Provider: "modal"})
	a.SnapshotStart()
	cost, signals := a.SnapshotEndAndBuild(1000)
	if cost != nil || signals != nil {
		t.Errorf("no NVML should produce no events; got %v / %v", cost, signals)
	}
}

// Decision #3 sharpening: sm_util_pct is task-window-averaged, NOT a point sample.
func TestGpuAccountantSmUtilPctWindowAveraged(t *testing.T) {
	resetNVMLForTests()
	mock := gpuMockWithDevices(1)
	SetNVMLBackendForTests(mock.AsBackend())
	t.Cleanup(resetNVMLForTests)

	a := NewGpuAccountant(GpuRuntimeModal, cloud.CloudEnv{Provider: "modal"})
	a.SnapshotStart()
	// duration_ms = 2000 — task ran 2 seconds, but only 1s of GPU activity
	// (TimeStamp delta from 0 → 1_000_000 µs = 1s). So sm_util_pct should
	// be ~50%, NOT 80% (the point-sample value).
	_, signals := a.SnapshotEndAndBuild(2000)
	if len(signals) != 1 {
		t.Fatalf("expected 1 signal; got %d", len(signals))
	}
	sig := signals[0]
	smUtilV, ok := sig["sm_util_pct"]
	if !ok || smUtilV == nil {
		t.Fatalf("sm_util_pct should be set; got %v", smUtilV)
	}
	smUtil, ok := smUtilV.(float64)
	if !ok {
		t.Fatalf("sm_util_pct should be float64; got %T", smUtilV)
	}
	if smUtil < 49 || smUtil > 51 {
		t.Errorf("Decision #3 — sm_util_pct should be window-averaged (~50%%); "+
			"got %v (point-sample bug?)", smUtil)
	}
}

// Decision #3 — sub-100ms degenerate window → sm_util_pct=nil.
func TestGpuAccountantDegenerateWindowEmitsNilSmUtilPct(t *testing.T) {
	resetNVMLForTests()
	mock := &MockNVMLBackend{
		Available:    true,
		DeviceCount:  1,
		ProductNames: map[int]string{0: "NVIDIA H100 80GB HBM3"},
		Memory:       map[int]NVMLMemInfo{0: {TotalBytes: 80 * 1024 * 1024 * 1024}},
		// No utilization samples at all — degenerate window.
		PerCallUtilization: map[int][]map[int]NVMLUtilSample{
			0: {{}, {}},
		},
	}
	SetNVMLBackendForTests(mock.AsBackend())
	t.Cleanup(resetNVMLForTests)

	a := NewGpuAccountant(GpuRuntimeModal, cloud.CloudEnv{Provider: "modal"})
	a.SnapshotStart()
	_, signals := a.SnapshotEndAndBuild(0) // 0-duration → degenerate
	if len(signals) != 1 {
		t.Fatalf("degenerate window should still emit one signal per device; got %d", len(signals))
	}
	sm := signals[0]["sm_util_pct"]
	if sm != nil {
		t.Errorf("degenerate sm_util_pct should be nil; got %v", sm)
	}
}

// ─── Decision #2 — MIG detected, full-billing applied ──────────────────

func TestGpuAccountantDetectsMIGAndEmitsTransparencyEvent(t *testing.T) {
	resetNVMLForTests()
	mock := &MockNVMLBackend{
		Available:    true,
		DeviceCount:  1,
		ProductNames: map[int]string{0: "NVIDIA H100 80GB HBM3"},
		MIGModes:     map[int]bool{0: true},
		Memory:       map[int]NVMLMemInfo{0: {TotalBytes: 80 * 1024 * 1024 * 1024}},
	}
	SetNVMLBackendForTests(mock.AsBackend())
	t.Cleanup(resetNVMLForTests)

	a := NewGpuAccountant(GpuRuntimeModal, cloud.CloudEnv{Provider: "modal"})
	a.SnapshotStart()
	cost, _ := a.SnapshotEndAndBuild(1000)
	if cost == nil {
		t.Fatalf("MIG-detected task should still emit cost event for transparency")
	}
	if cost["mig_profile"] == nil {
		t.Errorf("Decision #2 — MIG should populate mig_profile; got nil")
	}
}

// ─── Registry pattern (Go SDK chose this over Python's task field) ─────

func TestGpuAccountantRegistryRoundtrip(t *testing.T) {
	ResetGpuAccountantRegistryForTests()
	t.Cleanup(ResetGpuAccountantRegistryForTests)

	a := NewGpuAccountant(GpuRuntimeModal, cloud.CloudEnv{Provider: "modal"})
	taskID := "task-abc"
	RegisterGpuAccountant(taskID, a)
	got := GetGpuAccountant(taskID)
	if got == nil {
		t.Fatalf("registered accountant should be retrievable")
	}
	removed := UnregisterGpuAccountant(taskID)
	if removed == nil {
		t.Errorf("unregister should return the accountant")
	}
	if GetGpuAccountant(taskID) != nil {
		t.Errorf("post-unregister lookup should return nil")
	}
}

// Decision #1 — sets _cgroup_scope_fallback when scope is not container.
// We can't easily mock /proc/self/cgroup; verify the field plumbs through
// when the accountant is constructed with an explicit cgroup-scope override.
func TestGpuAccountantPropagatesCgroupFallbackLabel(t *testing.T) {
	resetNVMLForTests()
	mock := gpuMockWithDevices(1)
	SetNVMLBackendForTests(mock.AsBackend())
	t.Cleanup(resetNVMLForTests)

	a := NewGpuAccountant(GpuRuntimeModal, cloud.CloudEnv{Provider: "modal"})
	// Force a degraded scope so the fallback label is set.
	a.SetScopeForTests(CgroupScope{Kind: CgroupKindBareMetalUserSlice})
	a.SnapshotStart()
	cost, _ := a.SnapshotEndAndBuild(1000)
	if cost == nil {
		t.Fatalf("degraded scope should still emit a zero/observability event")
	}
	if cost["_cgroup_scope_fallback"] != "no_container_scope" {
		t.Errorf("expected fallback label no_container_scope; got %v", cost["_cgroup_scope_fallback"])
	}
}
